code/lib/core-server/src/utils/StoryIndexGenerator.ts

```diff
@@ -44,6 +44,8 @@ type ErrorEntry = {
 };
 type CacheEntry = false | StoriesCacheEntry | DocsCacheEntry | ErrorEntry;
 type SpecifierStoriesCache = Record<Path, CacheEntry>;
+// Extended type to keep track of the csf meta so we know the component id when referencing docs in `extractDocs`
+type StoryIndexEntryWithMeta = StoryIndexEntry & { meta?: any };
```

Comment:
Why not type `meta` also?

Reply:
Right! would `StoryIndexEntry & { meta?: { id?: string }}` be sufficient?

Reply:
Or how about `StoryIndexEntry & { meta?: IndexedCSFFile['meta'] };`?

Reply:
That sounds good. Also should we change the type of `StoriesCacheEntry` to have `entries: [null | StoryIndexEntryWithMeta | DocsIndexEntry]` to avoid the typecasts?

Reply:
Um, I'm actually uncertain about the full implications of modifying `StoriesCacheEntry['entries']` to include `StoryIndexEntryWithMeta`. Since it would affect `CacheEntry` and `SpecifierStoriesCache` types and the consumers of them.

I'd appreciate If you could give me some guidance as to what you'd like to see!

Oh and by the way `StoriesCacheEntry['entries']` can be `null` as well?

Reply:
No worries, thanks for being careful. I might be missing something but I believe the `SpecifierStoriesCache` is an internal type that's just used in this one file and not the return type of any public functions. So the implications should be confined to this type. Did I miss something?

> I'd appreciate If you could give me some guidance as to what you'd like to see!

As the type just represents the data we cache in this file to help with `ensureExtracted`, I think we can just add more fields to the type without problems.

> Oh and by the way StoriesCacheEntry['entries'] can be null as well?

Sorry I mean `CacheEntry`:

https://github.com/storybookjs/storybook/blob/29be84d00a4eae5597f8c71c074da32aa7cc7e42/code/lib/core-server/src/utils/StoryIndexGenerator.ts#L45

The cache entry is null when we haven't processed that particular file yet.

Reply:
So, yes, I was probably quite confusing there, I was typing from my phone, always a mistake. What I meant was let's make it:

```ts
type StoriesCacheEntry = {
  entries: (StoryIndexEntryWithMeta | DocsIndexEntry)[];
  dependents: Path[];
  type: 'stories';
};
type CacheEntry = false | StoriesCacheEntry | DocsCacheEntry | ErrorEntry;
type SpecifierStoriesCache = Record<Path, CacheEntry>;

```

Reply:
I think the `SpecifierStoriesCache` is only used by the `ensureExtracted` method and its submethods (`extractStories` and `extractDocs`) so if we want to add some more fields to help them communicate, that's fine with me!

Reply:
Okay great! I'll work on a commit that reflects what we've discussed in this thread. Thanks again for the detailed feedbacks 🙏.

I have one more question though. I've realized when I switch out `StoryIndexEntry` with `StoryIndexEntryWithMeta` in `StoriesCacheEntry['entries']` and remove the typecast in `extractStories` below (the last line):

```ts
 entries.push({
          id,
          title: csf.meta.title,
          name,
          importPath,
          tags,
          type: 'story',
          // We need to keep track of the csf meta so we know the component id when referencing docs below in `extractDocs`
          meta: csf.meta,
        } as StoryIndexEntryWithMeta);
```

Then I'll get a TypeScript error ('meta' does not exist in type 'StoryIndexEntry') because at the beginning of the body of `extractStories` we have `const entries = [] as IndexEntry[];`, and `IndexEntry` is `StoryIndexEntry | DocsIndexEntry`. 

This got me thinking whether if we should remove the entire usage of `IndexEntry` in `StoryIndexGenerator` for the sake of consistency? (so for example, should we create a similar internal type `type IndexEntryWithMeta = StoryIndexEntryWithMeta | DocsIndexEntry` and use that type instead throughout the file?)

Reply:
Would the return type of `ensureExtracted()` not still need to reference `IndexEntry` @sookmax?

Reply:
Yes it does! thanks I apparently forgot about it 😅. 

Then, should I just change `const entries = [] as IndexEntry[];` to `const entries = [] as (StoryIndexEntryWithMeta | DocsIndexEntry)[];` in `extractStories()`?

Reply:
@tmeasday I've just pushed a93d7b4227441855d45a5e3e4d91a0f9406e233f. Could you look at it to see if everything's okay? Sorry I bombarded you with lots of minor questions!